addon/components/connect.js

```diff
@@ -1,98 +1,113 @@
+    set: () => assert(`Cannot set redux property "${key}".`)
+  });
+}
+
+/**
+  Return an object of attrs passed to this Component.
+
+  `Component.attrs` is an object that can look like this:
+
+    {
+      myAttr: {
+        value: 'myValue'
+      }
+    }
 
-  if (!stateToComputed) {
-    stateToComputed = () => ({});
-  }
+  Ember provides that a `get` will return the value:
 
-  return Component => {
+    this.get('myAttr') === 'myValue'
 
-    return Component.extend({
+  @method getAttrs
+  @return {Object} an object of key, value for each attr
+  @private
+*/
+function getAttrs(context) {
+  const attrKeys = Object.keys(context.attrs || {});
+  return getProperties(context, attrKeys);
+}
 
+export default (stateToComputed, dispatchToActions) => {
+  return Component => {
+    return Component.extend({
       redux: service(),
 
       init() {
         const redux = this.get('redux');
 
-        let props = stateToComputed.call(this, redux.getState(), this.getAttrs());
+        if (stateToComputed) {
+          const getState = () => stateToComputed.call(this, redux.getState(), getAttrs(this));
 
-        Object.keys(props).forEach(name => {
-          defineProperty(this, name, computed(() =>
-            stateToComputed.call(this, redux.getState(), this.getAttrs())[name]
-          ).property().readOnly());
-        });
+          let state = getState();
 
-        if (!isEmpty(Object.keys(props))) {
-          this.unsubscribe = redux.subscribe(() => {
-            this.handleChange();
+          Object.keys(state).forEach(key => {
+            defineProperty(this, key, computedReduxProperty(key, () => state));
           });
-        }
 
-        this.actions = Object.assign({},
-          this.actions, dispatchToActions.call(this, redux.dispatch.bind(redux))
-        );
+          const handleChange = () => {
+            const newState = getState();
+            if (newState === state) return;
 
-        this._super(...arguments);
-      },
+            const propsToNotify = changedKeys(state, newState);
 
-      handleChange() {
-        const redux = this.get('redux');
+            state = newState;
 
-        const props = stateToComputed.call(this, redux.getState(), this.getAttrs());
+            if (propsToNotify.length > 0) {
+              run.join(() => {
+                propsToNotify.forEach(key => this.notifyPropertyChange(key));
+              });
+            }
+          };
 
-        const notifyProperties = Object.keys(props).filter(name => {
-          return this.get(name) !== props[name];
-        });
+          const unsubscribe = redux.subscribe(handleChange);
 
-        if (notifyProperties.length > 0) {
-          run.join(() => {
-            notifyProperties.forEach(name => this.notifyPropertyChange(name));
+          this.on('didUpdateAttrs', handleChange);
+          this.one('willDestroyElement', () => {
+            this.off('didReceiveAttrs', handleChange);
+            unsubscribe();
```

Comment:
I'll start with the cons:
* Less explicit that we're relying on this hook.
* No control of the sequence with respect to the superclass invocation (`this._super`).

The pro is that by defining everything within our `init` function, we can take advantage of the closure and keep all of our state local and internal to `ember-redux`, without exposing any internal data (like the `state` property and the `unsubscribe` method).

There are two alternatives I can think of right now:
1. Keep our existing implementation where we define state on the component instance itself. In #41 @toranb tried this, but it seems like the `_props` key he used might have conflicted with another addon. We could use something more namedspaced like `__ember_redux_state__` as a key instead.
2. Use the component `guid` (from Ember) and store this state in a local, global hash from within ember-redux, like:
```
const subscriptionForComponent = {}
...
export default (stateToComputed, dispatchToActions) => {
  return Component => {
    return Component.extend({
      ...
      willDestroy() {
        this._super(...arguments);
        if (subscriptionForComponent[guidFor(this)]) {
          subscriptionForComponent[guidFor(this)]();
          delete subscriptionForComponent[guidFor(this)];
        }
      }
...
```

I actually rather like the second, but I'm not sure if there's any more general commentary or literature about keeping local state within an es module.